### PR TITLE
IGNITE-21234 Introduce a more flexible way of triggering "too many dirty pages" checkpoints

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-21234

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java [hunk 1]
```diff
@@ -1291,6 +1305,28 @@ public Set<FullPageId> dirtyPages() {
         return res;
     }
 
+    /**
+     * Checkpoint urgency status enum.
+     */
+    public enum CheckpointUrgency {
```
Comment (tkalkirill) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java:1311]: Maybe it should be in a separate file?

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java [hunk 2]
```diff
@@ -1291,6 +1305,28 @@ public Set<FullPageId> dirtyPages() {
         return res;
     }
 
+    /**
+     * Checkpoint urgency status enum.
+     */
```
Comment (tkalkirill) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java:1310]:
```suggestion
    /** Checkpoint urgency status enum. */
```

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java [hunk 3]
```diff
@@ -1291,6 +1305,28 @@ public Set<FullPageId> dirtyPages() {
...
+    public enum CheckpointUrgency {
+        /**
+         * Signifies that there's enough free space in page memory and there's no need to trigger checkpoint.
```
Comment (tkalkirill) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java:1313]: What does “enough” mean? Please expand on this concept.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java [hunk 4]
```diff
@@ -1381,7 +1420,8 @@ private Segment(int idx, DirectMemoryRegion region) {
                     pool.pages()
             );
 
-            maxDirtyPages = pool.pages() * 3L / 4;
+            dirtyPagesSoftThreshold = pool.pages() * 3L / 4;
```
Comment (tkalkirill) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java:1423]: Is there a justification for choosing these values?

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java (hunk 1 above)
Comment (ibessonov) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java:1311]: Why not, I'll do it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java (hunk 4 above)
Comment (ibessonov) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java:1423]: No, these are educated guesses right now

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java (hunk 3 above)
Comment (ibessonov) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java]: Fixed

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java (hunk 4 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PersistentPageMemory.java:1401]: I don’t like this answer, but I can’t really object if we correct it later.

**Pull request merged**